src/DIRAC/WorkloadManagementSystem/Service/VirtualMachineManagerHandler.py

```diff
@@ -0,0 +1,479 @@
+        gLogger.error('haltInstances: on createEndpoint call: %s' % result['Message'])
+        continue
+
+      endpoint = result['Value']
+
+      # Get proxy to be used to connect to the cloud endpoint
+      authType = endpoint.parameters.get('Auth')
+      if authType and authType.lower() in ['x509', 'voms']:
+        siteName = endpoint.parameters['Site']
+        ceName = endpoint.parameters['CEName']
+        gLogger.verbose("Getting cloud proxy for %s/%s" % (siteName, ceName))
+        result = getProxyFileForCloud(endpoint)
+        if not result['OK']:
+          continue
+        endpoint.setProxy(result['Value'])
+
+      result = endpoint.stopVM(uniqueID)
+      if result['OK']:
+        cls.virtualMachineDB.recordDBHalt(instanceID, 0)
+        successful[instanceID] = True
+      else:
+        failed[instanceID] = result['Message']
+
+    return S_OK({"Successful": successful, "Failed": failed})
+
+  @classmethod
+  def getPilotOutput(cls, pilotRef):
+    if not pilotRef.startswith('vm://'):
+      return S_ERROR('Invalid pilot reference %s' % pilotRef)
+
+    # Get the VM public IP
+    diracID, nPilot = os.path.basename(pilotRef).split(':')
+    result = cls.virtualMachineDB.getUniqueIDByName(diracID)
+    if not result['OK']:
+      return result
+    uniqueID = result['Value']
+    result = cls.virtualMachineDB.getInstanceID(uniqueID)
+    if not result['OK']:
+      return result
+    instanceID = result['Value']
+    result = cls.virtualMachineDB.getInstanceParameter("PublicIP", instanceID)
+    if not result['OK']:
+      return result
+    publicIP = result['Value']
+
+    op = Operations()
+    privateKeyFile = op.getValue('/Cloud/PrivateKey', '')
+    diracUser = op.getValue('/Cloud/VMUser', '')
+
+    ssh_str = '%s@%s' % (diracUser, publicIP)
+    cmd = ['ssh', '-i', privateKeyFile, ssh_str,
+           "cat /etc/joboutputs/vm-pilot.%s.log" % nPilot]
+    inst = Popen(cmd, stdout=PIPE, stderr=PIPE, stdin=PIPE)
+    output, stderr = inst.communicate()
+    if inst.returncode:
+      return S_ERROR('Failed to get pilot output: %s' % stderr)
+    else:
+      return S_OK(output)
+
+  def initialize(self):
+    credDict = self.getRemoteCredentials()
+    self.rpcProperties = credDict['properties']
+
+  types_getCEInstances = [(list, type(None)), (list, type(None)), six.string_types]
+
+  def export_getCEInstances(self, siteList, ceList, vo):
+
+    if not siteList:
+      siteList = None
+    return self.getCEInstances(siteList=siteList, ceList=ceList, vo=vo)
+
+  types_stopInstance = [six.string_types, six.string_types, six.string_types]
+
+  def export_stopInstance(self, site, endpoint, nodeID):
+
+    return self.stopInstance(site, endpoint, nodeID)
+
+  types_getPilotOutput = [six.string_types]
+
+  def export_getPilotOutput(self, pilotReference):
+
+    return self.getPilotOutput(pilotReference)
+
+  types_checkVmWebOperation = [six.string_types]
+
+  def export_checkVmWebOperation(self, operation):
```

Comment:
Why is this needed?

Reply:
This is used by the web interface to change whether the admin functions are visibly enabled/disabled.